### PR TITLE
Vectorize

### PR DESCRIPTION
**Pull request by piepie62**
Switches from using the max values to using sets of available values for moves, abilities, species, items, and balls.

### CODE REVIEW

File: 3ds/source/gui/overlay/MoveOverlay.cpp [hunk 1]
```diff
@@ -66,11 +66,14 @@ MoveOverlay::MoveOverlay(ReplaceableScreen& screen, const std::variant<std::shar
 {
     instructions.addBox(false, 75, 30, 170, 23, COLOR_GREY, i18n::localize("SEARCH"), COLOR_WHITE);
     const std::vector<std::string>& rawMoves = i18n::rawMoves(Configuration::getInstance().language());
-    for (int i = 1; i <= TitleLoader::save->maxMove(); i++)
+    const std::set<int>& availableMoves      = TitleLoader::save->availableMoves();
+    for (auto i = availableMoves.begin(); i != availableMoves.end(); i++)
     {
-        if (i >= 622 && i <= 658)
+        if (*i >= 622 && *i <= 658)
```
Comment (Admiral-Fish) [inline 3ds/source/gui/overlay/MoveOverlay.cpp]: Could combine the if statements

File: 3ds/source/gui/overlay/PkmItemOverlay.cpp [hunk 2]
```diff
@@ -70,15 +70,18 @@ PkmItemOverlay::PkmItemOverlay(ReplaceableScreen& screen, std::shared_ptr<PKX> p
...
     const std::vector<std::string>& rawItems = i18n::rawItems(Configuration::getInstance().language());
-    for (int i = 1; i <= TitleLoader::save->maxItem(); i++)
+    const std::set<int>& availableItems      = TitleLoader::save->availableItems();
+    for (auto i = availableItems.begin(); i != availableItems.end(); i++)
     {
-        if (rawItems[i].find("\uFF1F\uFF1F\uFF1F") != std::string::npos || rawItems[i].find("???") != std::string::npos)
+        if (*i == 0)
             continue;
-        else if (i >= 807 && i <= 835)
+        else if (rawItems[*i].find("\uFF1F\uFF1F\uFF1F") != std::string::npos || rawItems[*i].find("???") != std::string::npos)
```
Comment (Admiral-Fish) [inline 3ds/source/gui/overlay/PkmItemOverlay.cpp]: Could there be a possible out of bounds here?

Comment (piepie62) [reply 3ds/source/gui/overlay/PkmItemOverlay.cpp]: On a possible out of bounds: Sure, it's possible. However, that would only happen in the case that either A) the save's availableItems is configured incorrectly or B) we don't have strings for some moves

File: 3ds/source/gui/overlay/MoveOverlay.cpp (hunk 1 above)
Comment (piepie62) [reply 3ds/source/gui/overlay/MoveOverlay.cpp]: Actually decided to handle this in a different way; I added a case in the if to sort 0 forwards

File: 3ds/source/gui/overlay/PkmItemOverlay.cpp (hunk 2 above)
Comment (piepie62) [reply 3ds/source/gui/overlay/PkmItemOverlay.cpp]: I did combine the ifs, though

**Comment by piepie62** [comment]
On memory impact: this will add a little bit more of a memory overhead, but nearly all of the value caching that's done in screens was there before this. I can't really think of any good way of removing the memory overhead and not hacking in a bunch of generation-specific logic (as was done with LGPE). I also used sets instead of vectors because they're more easily mutable and will not allocate blocks of memory for no reason

File: core/source/sav/Sav4.cpp [hunk 3]
```diff
@@ -893,3 +893,68 @@ std::string Sav4::pouchName(Pouch pouch) const
             return "";
     }
 }
+
+const std::set<int>& Sav4::availableItems(void) const
+{
+    static std::set<int> ret;
```
Comment (BernardoGiordano) [inline core/source/sav/Sav4.cpp]: repetition can be avoided by having just one function that incorporates this body and gets start and end values

Comment (piepie62) [reply core/source/sav/Sav4.cpp]: Good point; doing that. The sets will remain static, though; I don't want to recompute them or return a copy every time

**Pull request merged**